# Deep Learning Regression with Admissions Data

## 1. Project Goal

#### In this project, I will develope a deep learning regression that predicts the probability (0 to 1) that a students will be accepted to graduate schools based on application factors 

In [19]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import InputLayer
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
    

## 2. Dataset 

#### The dataset is from [Kaggle](https://www.kaggle.com/mohansacharya/graduate-admissions?select=Admission_Predict_Ver1.1.csv) containing information about 500 applicants from various universities and their chance of getting admitted

In [2]:
admission = pd.read_csv('admissions_data.csv')

In [3]:
admission

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
0             1        337          118                  4  4.5   4.5  9.65   
1             2        324          107                  4  4.0   4.5  8.87   
2             3        316          104                  3  3.0   3.5  8.00   
3             4        322          110                  3  3.5   2.5  8.67   
4             5        314          103                  2  2.0   3.0  8.21   
..          ...        ...          ...                ...  ...   ...   ...   
495         496        332          108                  5  4.5   4.0  9.02   
496         497        337          117                  5  5.0   5.0  9.87   
497         498        330          120                  5  4.5   5.0  9.56   
498         499        312          103                  4  4.0   5.0  8.43   
499         500        327          113                  4  4.5   4.5  9.04   

     Research  Chance of Admit   
0           1              0.92  
1           1              0.76  
2           1              0.72  
3           1              0.80  
4           0              0.65  
..        ...               ...  
495         1              0.87  
496         1              0.96  
497         1              0.93  
498         0              0.73  
499         0              0.84  

[500 rows x 9 columns]

The Dataset includes nine different columns as below:

* Serial No.: Index of each row (1-500)
* GRE Score: GRE test score (out of 340)
* TOEFL Score: TOEFL test score (out of 120)
* University Rating: Evaluated university rating (out of 5)
* SOP: Statement of Purpose Strength (out of 5)
* LOR: Letter of Recommendation Strength (out of 5)
* CGPA: Undergraduate GPA (out of 10)
* Research: Has research experience (either 0 or 1)
* Chance of Admit: Applicant’s chance of being admitted (ranging from 0 to 1)

## 3. Implementing Deep Learning 

### a. Preprocessing Data for learning

 (1) Separating features from labels using array slicing

In [4]:
#drop the first column (serial No.)
admission = admission.drop(columns = ['Serial No.'])

#Choose all columns except for last column for features
features = admission.iloc[:,0:-1]

#Choose the last column for labels
labels = admission.iloc[:,-1]

 (2) Splitting the data into training and test sets (test_size = 0.33)

In [5]:
#split the data into training and test sets 
feature_train,feature_test,label_train,label_test = train_test_split(features,labels,test_size=0.33, random_state =42)

 (3) Standardize the numerical features 

In [6]:
#standardize the numeric columns using Columntransformer
ct = ColumnTransformer([('standardize', StandardScaler(),[True,True,True,True,True,True,True])])
features_train_scale=ct.fit_transform(feature_train)
features_test_scale = ct.fit_transform(feature_test)

### b. Designing a Sequential model 

 (1) Creating input, output and hidden layers  

In [7]:
#create a Sequential object
model = Sequential()

#create input layers
input = InputLayer(input_shape = (features.shape[1],))

#add the input layer
model.add(input)

#add a hidden layer with 200 neurons
model.add(Dense(200,activation = 'relu'))

#add an output layer
model.add(Dense(1))

 (2) Choosing a learning rate hyperparameter using an optimizer: Adam

In [8]:
opt = Adam(learning_rate = .01)
model.compile(loss='mse',metrics=['mae'],optimizer=opt)

### c. Training the Model

In [12]:
#Train the model
model.fit(features_train_scale,label_train,epochs=40,batch_size=1,verbose=1)

Epoch 1/40
335/335 [==============================] - 0s 527us/step - loss: 0.0878 - mae: 0.1800
Epoch 2/40
335/335 [==============================] - 0s 504us/step - loss: 0.0113 - mae: 0.0816
Epoch 3/40
335/335 [==============================] - 0s 502us/step - loss: 0.0103 - mae: 0.0777
Epoch 4/40
335/335 [==============================] - 0s 510us/step - loss: 0.0068 - mae: 0.0628
Epoch 5/40
335/335 [==============================] - 0s 512us/step - loss: 0.0057 - mae: 0.0578
Epoch 6/40
335/335 [==============================] - 0s 498us/step - loss: 0.0065 - mae: 0.0632
Epoch 7/40
335/335 [==============================] - 0s 488us/step - loss: 0.0053 - mae: 0.0557
Epoch 8/40
335/335 [==============================] - 0s 486us/step - loss: 0.0057 - mae: 0.0565
Epoch 9/40
335/335 [==============================] - 0s 483us/step - loss: 0.0052 - mae: 0.0541
Epoch 10/40
335/335 [==============================] - 0s 484us/step - loss: 0.0063 - mae: 0.0588
Epoch 11/40
335/335 [========

In [14]:
#evaluate the model
val_mse, val_mae = model.evaluate(features_test_scale,label_test,verbose=0)

In [16]:
print('MAE: ', val_mae)

MAE:  0.04948580265045166


In [17]:
print('MSE:', val_mse)

MSE: 0.0050109680742025375
